### ISSUE TITLE
PyTorch vs jammi: BioClinical-ModernBERT LoRA triplet training and attention

### ISSUE THREAD

**Issue by vijaydevmasters** [f-inverse/jammi-ai#350]
## Context

Same model (`thomas-sounack/BioClinical-ModernBERT-large`), same task (LoRA on a triplet embedding objective). Two stacks:

- **PyTorch:** `nlp-riskadj-ml-jobs/train_embedding_model.py` (HuggingFace `AutoModel` + PEFT)
- **jammi:** encoder-adapter fine-tune (`target_modules` non-empty) in `jammi-ai` (Candle `ModernBert` + `jammi_lora`)

This issue records how they differ, and how attention is computed in each. No change is proposed here.

---

## 1. What both do (common ground)

Both train **LoRA**, not full fine-tune and not QLoRA. Frozen backbone weights stay put; trainable tensors are low-rank A/B on selected linears.

Both use **cosine-margin triplet loss** on L2-normalized embeddings:

```text
L = mean(relu(margin - cos(anchor, positive) + cos(anchor, negative)))
  = mean(max(0, cos(anchor, negative) - cos(anchor, positive) + margin))

margin = 0.3
```

PyTorch: `F.relu(margin - pos_sim + neg_sim).mean()`.  
jammi: `max(0, neg_sim - pos_sim + margin).mean_all()` in `TrainingLoop::triplet_loss`.

Both encode a batch as **three forwards** (anchors, then positives, then negatives), mean-pool token states (mask out pads), L2-normalize, then the loss above. Then reverse-mode autodiff into LoRA A/B only, gradient clip (global L2, max_norm=1.0), AdamW.

Worked shapes for BioClinical-ModernBERT-large, batch 8, sequence length 128:

```text
hidden     [8, 128, 1024]
Wqkv out   [8, 128, 3072]     # fused Q|K|V
Q, K, V    [8, 16, 128, 64]   # 16 heads, head_dim 64
scores     [8, 16, 128, 128]  # 2,097,152 scores per layer
FFN Wi     [8, 128, 5248]     # 2 × 2624 (gate + up)
after FFN  [8, 128, 1024]
pooled     [8, 1024]          # after all layers + mean pool + L2
```

---

### Backbone / attention pattern

| | PyTorch script | jammi |
|---|---|---|
| Model load | `AutoModel.from_pretrained(..., trust_remote_code=True)` — official HuggingFace ModernBERT | Candle port `jammi_encoders::ModernBert` from `config.json` + `model.safetensors` |
| Local vs global attention | **Honored.** Layers 0, 3, 6, … global (full sequence). Other layers **local** sliding window (`local_attention=128`, typically 64 tokens each side) | **Not honored.** Config fields are parsed then ignored. **Every layer is global** (padding mask only) |
| RoPE | Global layers: `rope_theta=160000`. Local layers: `rope_theta=10000` | **One** RoPE table per layer, always `global_rope_theta` (160000). `local_rope_theta` unused |
| Position encoding | RoPE on Q and K (HF) | RoPE on Q and K (jammi), same rotate-half formula, wrong θ on layers that HF treats as local |

jammi comment on `ModernBertConfig`: sliding-window fields are accepted for load compatibility but **not exercised**. Exact vs HF only if `global_attn_every_n_layers == 1` (every layer global). BioClinical-ModernBERT-large is **not** that checkpoint (`global_attn_every_n_layers=3`).

The 1-layer `tiny_modernbert` fixture cannot show this gap: layer 0 is global in both stacks.

### LoRA sites

| | PyTorch | jammi |
|---|---|---|
| Targets (default / typical) | `['Wqkv', 'out_proj', 'Wi', 'Wo']` | `target_modules` e.g. `Wqkv`, `Wo`, `Wi`; `Wo` matches **both** `attn.Wo` and `mlp.Wo` |
| `out_proj` | BERT/RoBERTa name; on ModernBERT this module usually **does not exist** (PEFT skips unmatched names) | N/A |
| Init | PEFT defaults (ZerosB-style) | `ZerosB`: A Kaiming-uniform, B zeros; scaling `alpha/rank` unless RSLoRA |
| Rank / alpha / dropout (script vs jammi defaults) | r=16, α=32, dropout=0.1 | r=8, α=16, dropout=0.05 |
| Empty `target_modules` | (script always uses PEFT on encoder) | Trains a **projection head** on frozen pooled embeddings; backbone forward is detached (`Tensor::from_vec` of inference floats) |

### Data / batching

| | PyTorch | jammi |
|---|---|---|
| Input | JSONL triplets | Parquet/CSV/etc. via DataFusion; columns `anchor`, `positive`, `negative` |
| Batch size (defaults) | 16 | 8 |
| Sequence length | `max_length=128` | `max_seq_length=512` |
| Padding | `padding=True` → **longest in this batch**, not always 128 | **Batch-longest** as well (`PaddingStrategy::BatchLongest`) |
| Truncation | yes, at 128 | yes, at `min(max_seq_length, encoder.max_seq_length())` |
| Shuffle | `DataLoader(shuffle=True)` | **Source order** (CPU fine-tune is a pure function of `(seed, source rows, config)`) |
| Workers / pin_memory | `num_workers=2`, `pin_memory=True` | no PyTorch DataLoader |

128 vs 512 is a **truncation cap**, not “PyTorch does not pad.” Short notes are padded to the longest example in that batch in both stacks.

### Loss / pooling

| | PyTorch | jammi |
|---|---|---|
| Pool | mean over masked tokens **inside** `EmbeddingModel` | mean pool **inside** encoder `forward`, then L2 |
| L2 | `F.normalize` **after** the three forwards | L2 inside `pool_and_normalize` |
| Triplet vs MNRL | script is always triplet-margin | triplet-margin unless `embedding_loss=mnrl` (then InfoNCE; explicit negatives become extra columns) |

Algebraically mean-pool + L2 is the same; AMP may compute PyTorch’s pool in FP32 (`mask.float()`).

### Optimizer / schedule

| | PyTorch | jammi |
|---|---|---|
| Optimizer | `torch.optim.AdamW` on `requires_grad` params | jammi-owned AdamW (same update; serializable moments for resume) |
| Learning rate | **2e-5** | **2e-4** |
| Warmup | **10% of total steps** = `0.1 * len(train_loader) * epochs` | **fixed 100 steps**, independent of dataset size |
| After warmup | cosine over remaining steps | cosine (or linear/constant) over remaining steps |
| Weight decay | 0.01 | 0.01 |
| Grad clip | 1.0 | 1.0 |
| Grad accumulation | config field `1`, unused | `gradient_accumulation_steps` used; backward each micro-batch, step every N |
| Scheduler step | every batch | every optimizer step |

Example: 10 000 triplets, batch 16, 10 epochs → PyTorch warmup ≈ 625 steps. Same data on jammi defaults is still 100 warmup steps.

### Precision

| | PyTorch | jammi |
|---|---|---|
| Weight load | `torch.float16` if `use_fp16` | mmap; `backbone_dtype` F32 default, optional F16/BF16 |
| Forward GEMMs | **AMP autocast:** Linear/attention GEMM in **FP16** (Tensor Cores); softmax / LayerNorm / loss typically **FP32** | Default **all F32**. Optional: frozen W in F16/BF16, **LoRA A/B always F32** |
| Loss scaling | **`GradScaler`**: `scale * loss` → backward → unscale → clip → AdamW; skip step on inf/NaN | **None** |
| BF16 | script uses FP16 AMP (correct for T4, which has no BF16) | BF16 is a GPU-tier compute precision; CPU load refuses it |

AMP is mixed precision **per op**, not “the whole graph is FP16.” GradScaler exists because FP16 gradients can underflow to 0.

### Backward / autodiff

| | PyTorch | jammi |
|---|---|---|
| Tape | `torch.autograd`; each op has `grad_fn`; saved activations for backward | Candle `Tensor.backward()` over the primitive-op graph |
| What gets `.grad` | tensors with `requires_grad=True` (LoRA A/B) | `Var`s in the `VarMap` (LoRA A/B). Mmaped backbone is not a `Var` |
| Attention backward | fused: FlashAttention backward (A100) or mem-efficient recompute (T4) | differentiate `matmul` and `softmax` **separately**; full L×L kept for backward |
| LayerNorm backward | fused ATen kernel | **primitive composition** (`mean`, `sub`, `sqr`, `sqrt`, `div`, `mul`) so each piece has `bwd`. Candle fused `ops::layer_norm` is eval-only (and ModernBERT LN has **no bias**, so even eval stays on the slow path) |
| Dropout | PEFT dropout on LoRA path | seeded dropout stream (CPU run-to-run byte identity) |

### Output artifact

| | PyTorch | jammi |
|---|---|---|
| During train | PEFT adapter dir (`save_pretrained`) | `adapter.safetensors` + `adapter_config.json` |
| After train | also **`merge_and_unload`**: bake `W ← W + (α/r)BA` into a full HF folder | adapter stays separate; serving reloads base + adapter |

### GPU libraries (not the CUDA Toolkit)

The CUDA Toolkit is nvcc, cuBLAS (GEMM), cuDNN, cudart. **FlashAttention is not in the toolkit.** Installing CUDA does not give Candle FlashAttention.

| | PyTorch | jammi / Candle |
|---|---|---|
| GEMM | cuBLAS `cublasGemmEx` under AMP (FP16 Tensor Cores) | `contiguous_matmul` → Candle CUDA PTX / cuBLAS, **F32** by default; `contiguous()` copy before GEMM |
| Attention | SDPA dispatcher (below) | naive `QKᵀ` + softmax + `PV` only |
| Driver story | wheel ships SASS for T4 (sm_75) and A100 (sm_80) | Candle PTX JIT; jammi GPU floor CUDA **12.6** / driver **r560+** |

---

### The formula (both stacks)

For each head:

```text
Attention(Q, K, V) = softmax(Q Kᵀ / √d + mask) V
```

Q, K, V come from a **fused** `Wqkv` (one linear 1024 → 3072), split, reshape to heads `[batch, heads, seq, head_dim]`. RoPE rotates Q and K (not V). `Wo` mixes heads after concat. Residual add. Then GeGLU FFN (`Wi` packs gate+up, `gelu(gate)*up`, `mlp.Wo`) + residual.

**Multi-head** = 16 independent 64-d attentions, not one 1024-d attention. Scores for this example:

```text
[8, 16, 128, 64] @ [8, 16, 64, 128] → [8, 16, 128, 128]
```

Softmax is **per row** (each token’s weights over the 128 keys sum to 1). Pad positions get a large negative (`-10000`) so softmax ≈ 0 there.

**Local window (HF only):** on local layers, token i may only attend to a band around i (≈ 64 left + 64 right), **and** pads are still masked. At L=128 this is **not** full attention: token 0 cannot see token 127 on a local layer. Global layers mix the whole note. jammi never applies the band; every real token sees every real token on every layer.

### Three ways to implement the same formula

```text
                    stores full L×L in VRAM?     typical GPU
naive / math              YES                    any (jammi)
memory-efficient          NO (chunks; recompute backward)   T4, A100, …
FlashAttention            NO (SRAM tiles)        A100+ (sm_80)
```

**Naive (jammi, PyTorch `MATH` backend)**

Several GPU kernels, all on-device:

1. Write L×L to **GPU VRAM** (`Q @ K.T`)
2. Read it, add mask
3. Read it, softmax, write a second L×L
4. Read that, multiply by V

“Back and forth” is **GPU ALUs ↔ GPU VRAM**, not CPU ↔ GPU. CPU↔VRAM (PCIe) is once per batch when the batch is copied onto the card.

Fitting Q, K, V and the score sheet in VRAM only means **no OOM**. At L=128 the sheet is small (~8 MB/layer FP32) and already fits; naive is still more VRAM traffic and more launches than fused kernels.

**Memory-efficient (PyTorch on T4)**

Algorithm: Rabe & Staats (Google, 2021) — chunk the sequence, softmax a slice, discard, **recompute** in backward.  
CUDA: Meta **xFormers** CUTLASS backend (`EFFICIENT_ATTENTION` / `sdpa_mem_eff`). Runs on Turing (T4, sm_75). Forward can be similar cost; backward pays extra compute to save memory. Fused vs jammi’s separate `matmul`/`softmax` launches.

**FlashAttention (PyTorch on A100)**

Dao, Fu, Ermon, Rudra, Ré (Stanford, NeurIPS 2022); FA2: Tri Dao (ICLR 2024). Tile Q,K,V in **on-chip SRAM**, **online softmax** (Milakov & Gimelshein, NVIDIA, 2018). The L×L matrix never exists in VRAM. Faster **even when L×L would have fitted**. Requires Ampere+; **T4 does not get this kernel**.

### PyTorch’s interface (SDPA)

PyTorch 2.0 does not expose raw CUDA launches. One operator:

```python
torch.nn.functional.scaled_dot_product_attention(query, key, value)
```

C++ dispatcher picks:

| Backend | Kernel | Typical device |
|---|---|---|
| `FLASH_ATTENTION` | Dao FA, **vendored in libtorch** | A100/H100, FP16/BF16 |
| `EFFICIENT_ATTENTION` | xFormers/CUTLASS mem-efficient | T4 and anything FA rejects |
| `CUDNN_ATTENTION` | NVIDIA cuDNN fused SDPA (not Dao’s repo) | newer PyTorch |
| `MATH` | matmul + softmax | CPU, fallback |

Pin a backend with `torch.nn.attention.sdpa_kernel(SDPBackend.FLASH_ATTENTION)`.

Those FA kernels are **compiled into the PyTorch wheel**, not installed by the CUDA Toolkit.

Second, separate door: `pip install flash-attn` + HuggingFace `attn_implementation="flash_attention_2"` (Dao’s package). The training script’s default HF path is SDPA, not necessarily that extra package.

### What jammi calls

```text
contiguous_matmul(Q, Kᵀ) → add pad mask → softmax → contiguous_matmul(attn, V)
```

No SDPA, no `flash-attn`, no `candle-flash-attn` in this repo. cuBLAS may run the two GEMMs; softmax is Candle’s own kernel. Local-window sparsity is not implemented, so FLOPs stay Θ(L²) on every layer.

### Who built what

| Piece | Who |
|---|---|
| Online softmax | Milakov, Gimelshein (NVIDIA, 2018) |
| Chunk/recompute mem-efficient algorithm | Rabe, Staats (Google, 2021) |
| FlashAttention | Dao, Fu, Ermon, Rudra, Ré (Stanford, 2022) |
| FlashAttention-2 | Tri Dao (2023/24) |
| xFormers mem-efficient CUDA | Meta FAIR (Lefaudeux, Massa, et al.) |
| SDPA dispatcher | Meta PyTorch (2.0, 2023) |
| cuDNN fused attention | NVIDIA (library, not the CUDA Toolkit “base” set of cuBLAS-only) |

---

## 4. Side-by-side (attention + training)

| | PyTorch (`train_embedding_model.py`) | jammi encoder-adapter |
|---|---|---|
| ModernBERT local/global + dual RoPE | yes (HF) | no (all-global, one θ) |
| Attention kernel | SDPA: FA2 (A100) or mem-efficient (T4) | naive L×L matmul+softmax |
| Tensor Cores on GEMM | yes (AMP FP16) | no (F32 default) |
| Loss scaler | GradScaler | none |
| Warmup | 10% of steps | 100 steps |
| LR | 2e-5 | 2e-4 |
| Seq cap | 128 | 512 |
| Batch / shuffle | 16 / yes | 8 / no |
| LoRA r, α, dropout | 16, 32, 0.1 | 8, 16, 0.05 |
| CPU adapter bit-identical given seed | not the script’s contract | yes (`Device::Cpu`) |
| Merged full weights | yes | adapter files only |

Same triplet **loss**. Different **function** (attention pattern + RoPE), different **precision recipe**, different **GPU kernels**, different **hyperparameters**. Embeddings from the two trained adapters are not expected to match.

**Comment by thankrandomness** [comment]
The correctness half of this is now fixed in #353 — thank you for writing it up precisely enough to act on.

**Local/global attention and dual RoPE are implemented.** Verified against goldens dumped from HuggingFace `transformers`, not against a second port. Pre-fix real-token divergence was 0.055 max-abs, 500× the 1e-4 parity tolerance.

The fixture gap this issue identifies was the reason the defect survived: every ModernBERT fixture in the tree is `num_hidden_layers=1` / `global_attn_every_n_layers=1`, and layer 0 is global under ModernBERT's own rule, so none of them can express the divergence. `cookbook/fixtures/tiny_modernbert_local` (4 layers, `global_attn_every_n_layers=3`, ±8 band over 64 tokens) can, and its generator **asserts its own discriminating power on every regeneration** — at the default initialiser scale attention is nearly uniform and the dual-RoPE half was invisible at ~1e-5, below any usable tolerance. Current margins are 855× and 95×.

**Two corrections to the comparison table**, both measured:

- *"Attention kernel: naive L×L matmul+softmax"* is accurate, but Flash Attention cannot close it for training: `candle-flash-attn`'s `FlashAttn` implements `CustomOp3` with **no `bwd`**, so the default returns `BackwardNotSupported`. It is forward-only. Same for `ops::softmax_last_dim` (`apply_op1_no_bwd`) — and that one is worse, because it silently truncates the backward graph rather than erroring.
- Attention is not the dominant cost at the sequence lengths in question. Step time is flat from seq 64 to 256 and only goes superlinear past 512.

The measured gap and its actual cause — candle materialising every intermediate to HBM for want of a fusion layer — are in #352.

The deliberate divergences (source-order determinism vs shuffle, adapter-only vs `merge_and_unload`) remain deliberate. The hyperparameter differences are defaults, not defects, though the fixed 100-step warmup being independent of dataset size is worth revisiting separately.

**Comment by thankrandomness** [comment]
Closing as fully addressed. This issue was a comparison record ("no change is proposed here"), and every actionable delta it surfaced has since landed on main:

- **The correctness gap it exposed — local/global alternating attention + dual RoPE — was fixed in #353**, verified against goldens dumped from HuggingFace transformers (pre-fix divergence 0.055 max-abs, 500× the 1e-4 parity tolerance). The implementation is live in `crates/jammi-encoders/src/modernbert.rs` (`local_attention` / `global_attn_every_n_layers`), and the fixture gap this record identified is closed by `cookbook/fixtures/tiny_modernbert_local` (4 layers, `global_attn_every_n_layers=3`), whose generator asserts its own discriminating power on regeneration.
- The precision-recipe delta (bf16 inert on LoRA-targeted linears) was fixed in #353 as well; the kernel/SDPA and throughput deltas are the ongoing #352 program (mem-efficient attention, fused LoRA site) and the encoder ports are tracked in #356.

The remaining perf/kernel differences this table recorded are all tracked by open issues (#352, #356, #374), so this record has served its purpose.
